### PR TITLE
Added ParamIterator for the cpu backend

### PR DESCRIPTION
**Pull request by umar456**

### CODE REVIEW

File: src/backend/cpu/ParamIterator.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    // The iterator's stride
+    const af::dim4 stride;
+
+    // NOTE: This is not really the true coordinate of the iteration. It's
+    // values will go down as you move through the array.
+    std::array<dim_t, AF_MAX_DIMS> dim_index;
+
+    /// Calculates the iterator offsets. These are different from the original offsets
+    /// because they define the stride from the end of the last element in the previous
+    /// dimention to the first element on the next dimenstion.
```
Comment (syurkevi) [inline src/backend/cpu/ParamIterator.hpp]:
```suggestion
    /// dimension to the first element on the next dimension.
```

File: src/backend/cpu/ParamIterator.hpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    bool operator!=(const ParamIterator& other) const noexcept {
+        return other.ptr != ptr;
+    }
+
+    /// Advances the iterator
+    ParamIterator operator++() noexcept {
+        for(int i = 0; i < AF_MAX_DIMS; i++) {
+            dim_index[i]--;
+            ptr += stride[i];
+            if(dim_index[i]) return *this;
```
Comment (syurkevi) [inline src/backend/cpu/ParamIterator.hpp]: Fix

**Comment by 9prady9** [comment]
We should be able to test the iterator with subarray if we create a array and use `host` member function on it right ? @umar456

File: src/backend/cpu/ParamIterator.hpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+#include <iterator>
+
+namespace cpu {
+
+/// A Param iterator that iterates through a Param object
+template<typename T>
+class ParamIterator {
+    T* ptr;
+
+    // NOTE: This is not really the true coordinate of the iteration. It's
```
Comment (mlloreda) [inline src/backend/cpu/ParamIterator.hpp]: It's -> its

**Pull request merged**